File: nodes/array-sort/array-sort.process.py
from datetime import datetime

async def process(inputs, settings, config, nodeConfig):
  array = inputs.get("array", [])
  if not isinstance(array, list):
    array = [array]
    
  order = settings.get("order", "ascending")
  type_name = settings.get("type", "string")
  property_name = settings.get("property")
  case_sensitive = settings.get("case_sensitive", True)
  
  def get_value(item):
    value = item.get(property_name) if property_name and isinstance(item, dict) else item
    
    if value is None:
      return None
      
    if type_name == "number":
      try:
        return float(value)
      except (ValueError, TypeError):
        return 0
        
    elif type_name == "date":
      try:
        return datetime.fromisoformat(str(value)).timestamp()
      except (ValueError, TypeError):
        return 0
        
    else: # string
      value = str(value)
      return value if case_sensitive else value.lower()
  
  def compare(item):
    value = get_value(item)
    return (value is None, value)  # None values sort last
    
  reverse = order != "ascending"
  return {
    "array": sorted(array, key=compare, reverse=reverse)
  } 

File: sdks/nodejs/nodes/array-sort/array-sort.process.js
export default async ({inputs, settings, config}) => {

  const array = Array.isArray(inputs.array) ? inputs.array : [inputs.array];
  const order = settings.order || "ascending";
  const type = settings.type || "string";
  const property = settings.property;
  const caseSensitive = settings.case_sensitive !== false;

  const sorted = [...array].sort((a, b) => {
    let valueA = property ? (a && a[property]) : a;
    let valueB = property ? (b && b[property]) : b;

    // Handle null/undefined values
    if (valueA == null) return 1;
    if (valueB == null) return -1;
    
    switch (type) {
      case "number":
        valueA = Number(valueA);
        valueB = Number(valueB);
        break;
        
      case "date":
        valueA = new Date(valueA).getTime();
        valueB = new Date(valueB).getTime();
        break;
        
      default: // string
        valueA = String(valueA);
        valueB = String(valueB);
        if (!caseSensitive) {
          valueA = valueA.toLowerCase();
          valueB = valueB.toLowerCase();
        }
    }

    if (valueA < valueB) return order === "ascending" ? -1 : 1;
    if (valueA > valueB) return order === "ascending" ? 1 : -1;
    return 0;
  });

  return {
    array: sorted
  };
}; 